### PR TITLE
[GORDO-1571] historic pregel access

### PR DESCRIPTION
**Pull request by hkernbach**
### Scope & Purpose

Follow up on:
[GORDO-1481] Persist pregel run data (#18278)

This PR grants a client (HTTP, arangosh-client and arangosh-server) access to the pregel historic entries. 

- [x] :pizza: New feature

### Checklist

- [x] Tests
  - [x] **integration tests**
- [x] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...) This still needs to be done!

#### Related Information

- [x] GitHub issue / Jira ticket: [GORDO-1481]

[GORDO-1481]: https://arangodb.atlassian.net/browse/GORDO-1481?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[GORDO-1481]: https://arangodb.atlassian.net/browse/GORDO-1481?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: arangod/Pregel/StatusWriter/CollectionStatusWriter.h [hunk 1]
```diff
@@ -131,34 +170,59 @@ struct CollectionStatusWriter : StatusWriterInterface {
...
-                         {}));
+          if (data.has_value()) {
+            auto payload = inspection::serializeWithErrorT(data.value());
+            TRI_ASSERT(operation == OperationType::DELETE_DOCUMENT);
+            return handleOperationResult(
+                trx, options, transactionResult,
+                trx.remove(StaticStrings::PregelCollection, payload->slice(),
+                           {}));
+          } else {
+            OperationOptions peter;
```
Comment (hkernbach) [inline arangod/Pregel/StatusWriter/CollectionStatusWriter.h:204]:
```suggestion
```

File: arangod/V8Server/v8-pregel.cpp [hunk 2]
```diff
@@ -263,6 +265,138 @@ static void JS_PregelAQLResult(
...
+  if (argLength != 1 || (!args[0]->IsNumber() && !args[0]->IsString())) {
+    // TODO extend this for named graphs, use the Graph class
+    TRI_V8_THROW_EXCEPTION_USAGE("_pregelHistory(<executionNum>]");
+  }
+
+  // Read single history entry
+  auto executionNumber = arangodb::pregel::ExecutionNumber{
+      TRI_ObjectToUInt64(isolate, args[0], true)};
+
+  VPackBuilder builder;
```
Comment (hkernbach) [inline arangod/V8Server/v8-pregel.cpp:392]:
```suggestion
```

File: arangod/V8Server/v8-pregel.cpp [hunk 3]
```diff
@@ -263,6 +265,138 @@ static void JS_PregelAQLResult(
...
```
Comment (hkernbach) [inline arangod/V8Server/v8-pregel.cpp:325]:
```suggestion
```

File: arangod/V8Server/v8-pregel.cpp [hunk 4]
```diff
@@ -263,6 +265,138 @@ static void JS_PregelAQLResult(
...
+    // Read all pregel history entries
+    auto result = pregel.handleHistoryRequest(
+        vocbase, arangodb::rest::RequestType::DELETE_REQ, std::nullopt);
+    handlePregelHistoryV8Result(result);
+    return;
+  }
+
+  if (argLength != 1 || (!args[0]->IsNumber() && !args[0]->IsString())) {
+    // TODO extend this for named graphs, use the Graph class
+    TRI_V8_THROW_EXCEPTION_USAGE("_pregelHistory(<executionNum>]");
```
Comment (hkernbach) [inline arangod/V8Server/v8-pregel.cpp:385]:
```suggestion
    TRI_V8_THROW_EXCEPTION_USAGE("_pregelHistoryRemove(<executionNum>]");
```

File: arangod/Pregel/StatusWriter/CollectionStatusWriter.h [hunk 5]
```diff
@@ -131,34 +170,59 @@ struct CollectionStatusWriter : StatusWriterInterface {
...
+          auto payload = inspection::serializeWithErrorT(data.value());
+          return handleOperationResult(
+              trx, options, transactionResult,
+              trx.documentAsync(StaticStrings::PregelCollection,
+                                payload->slice(), {})
+                  .get());
+        } else {
+          size_t skip = 0;
+          size_t limit = 0;
+          return trx.all(StaticStrings::PregelCollection, skip, limit, options);
```
Comment (hkernbach) [inline arangod/Pregel/StatusWriter/CollectionStatusWriter.h:224]: This is not allowed to be called from a Coordinator and fails currently.

File: arangod/Pregel/StatusWriter/CollectionStatusWriter.h [hunk 6]
```diff
@@ -58,6 +58,17 @@ struct CollectionStatusWriter : StatusWriterInterface {
...
+    CollectionNameResolver resolver(_vocbaseGuard.database());
+    auto logicalCollection =
+        resolver.getCollection(StaticStrings::PregelCollection);
+    if (logicalCollection == nullptr) {
+      THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_ARANGO_DATA_SOURCE_NOT_FOUND,
+                                     StaticStrings::PregelCollection);
+    }
+    _logicalCollection = std::move(logicalCollection);
+  };
+
```
Comment (hkernbach) [inline arangod/Pregel/StatusWriter/CollectionStatusWriter.h:71]: As this header file does now include way too much code - we need to split this up into .cpp and .h files.

**Review by jvolmer — COMMENTED** [review]
I added several comments to improve readability of the code. In general I noticed that you added a lot of if-statements inside functions, which most of the time make the code harder to read.
The code gets much cleaner if you have several small building blocks that you combine for each of your cases separately than having a big and possibly nested if-statement that includes all cases. This makes it also easier to add a new case: You don't have to fiddle around in you if-statement but just write a new case that combines you already defined building blocks in a different way.
This applies to your Rest Handlers, CollectionStatusWriter and also your tests.

File: arangod/Pregel/StatusWriter/CollectionStatusWriter.h (hunk 6 above)
Comment (markuspf) [reply arangod/Pregel/StatusWriter/CollectionStatusWriter.h:72]: https://arangodb.atlassian.net/browse/GORDO-1592

File: arangod/Pregel/Conductor/Conductor.cpp [hunk 7]
```diff
@@ -748,31 +749,78 @@ void Conductor::toVelocyPack(VPackBuilder& result) const {
...
+  };
+
+  if (_state == ExecutionState::DONE) {
+    // TODO: What I wanted to do here is to just use the already available
+    // toVelocyPack() method. This fails currently because of the lock:
+    // "[void arangodb::Mutex::lock()]: _holder != Thread::currentThreadId()"
+    // Therefore, for now - do it manually. Let's clean this up ASAP.
+    // this->toVelocyPack(stateBuilder);
+    // After this works, we can remove all of that code below (same scope).
+    // Including those lambda helper methods.
```
[the remark this answers is not in the document]
Comment (markuspf) [reply arangod/Pregel/Conductor/Conductor.cpp:804]: Should be addressed as part of https://arangodb.atlassian.net/jira/software/c/projects/GORDO/boards/40?modal=detail&selectedIssue=GORDO-1584

File: arangod/Pregel/REST/RestControlPregelHandler.cpp [hunk 8]
```diff
@@ -152,33 +159,102 @@ void RestControlPregelHandler::getExecutionStatus() {
     return;
   }
 
-  if (suffixes.size() != 1 || suffixes[0].empty()) {
-    generateError(
-        rest::ResponseCode::BAD, TRI_ERROR_HTTP_SUPERFLUOUS_SUFFICES,
-        "superfluous parameter, expecting /_api/control_pregel[/<id>]");
+  if (suffixes.size() == 1 && suffixes.at(0) != "history") {
+    if (suffixes[0].empty()) {
```
[the remark this answers is not in the document]
Comment (markuspf) [reply arangod/Pregel/REST/RestControlPregelHandler.cpp:163]: best to do this with `lib/Rest/PathMatch.h` in a later PR, I think

File: arangod/Pregel/REST/RestControlPregelHandler.cpp [hunk 9]
```diff
@@ -152,33 +159,102 @@ void RestControlPregelHandler::getExecutionStatus() {
...
+      generateError(
+          rest::ResponseCode::BAD, TRI_ERROR_HTTP_SUPERFLUOUS_SUFFICES,
+          "superfluous parameter, expecting /_api/control_pregel[/<id>]");
```
[the remark this answers is not in the document]
Comment (hkernbach) [reply arangod/Pregel/REST/RestControlPregelHandler.cpp:166]: https://arangodb.atlassian.net/browse/GORDO-1593

File: arangod/Pregel/REST/RestControlPregelHandler.cpp [hunk 10]
```diff
@@ -152,33 +159,102 @@ void RestControlPregelHandler::getExecutionStatus() {
...
+      generateError(rest::ResponseCode::NOT_FOUND, TRI_ERROR_CURSOR_NOT_FOUND,
+                    "Execution number is invalid");
+      return;
+    }
+
+    VPackBuilder builder;
+    c->toVelocyPack(builder);
+    generateResult(rest::ResponseCode::OK, builder.slice());
     return;
+  } else if ((suffixes.size() >= 1 || suffixes.size() <= 2) &&
```
[the remark this answers is not in the document]
Comment (hkernbach) [reply arangod/Pregel/REST/RestControlPregelHandler.cpp:183]: https://arangodb.atlassian.net/browse/GORDO-1593

File: arangod/Pregel/PregelFeature.cpp [hunk 11]
```diff
@@ -1060,6 +1062,41 @@ void PregelFeature::handleWorkerRequest(TRI_vocbase_t& vocbase,
   }
 }
 
+ResultT<OperationResult> PregelFeature::handleHistoryRequest(
```
[the remark this answers is not in the document]
Comment (hkernbach) [reply arangod/Pregel/PregelFeature.cpp:1003]: https://arangodb.atlassian.net/browse/GORDO-1594

File: arangod/Pregel/REST/RestControlPregelHandler.cpp [hunk 12]
```diff
@@ -152,33 +159,102 @@ void RestControlPregelHandler::getExecutionStatus() {
...
+    generateResult(rest::ResponseCode::OK, VPackSlice::trueSlice());
+  }
 }
 
-void RestControlPregelHandler::cancelExecution() {
+void RestControlPregelHandler::handleDeleteRequest() {
   std::vector<std::string> const& suffixes = _request->decodedSuffixes();
+
+  if ((suffixes.size() >= 1 || suffixes.size() <= 2) &&
+      suffixes.at(0) == "history") {
```
[the remark this answers is not in the document]
Comment (hkernbach) [reply arangod/Pregel/REST/RestControlPregelHandler.cpp:242]: https://arangodb.atlassian.net/browse/GORDO-1593

File: tests/js/common/shell/shell-pregel-status-writer.js [hunk 13]
```diff
@@ -118,6 +216,124 @@ function pregelStatusWriterSuite() {
...
+
+      const nonNumerics = [true, false, "aString"];
+
+      nonNumerics.forEach(pid => {
+        verifyPersistedStatePIDNotAvailableRead(pid, false);
+      });
+    },
+
+    testReadAllHistoricEntriesHTTPAndDeleteAfterwards: function () {
+      if (isServer) {
```
[the remark this answers is not in the document]
Comment (hkernbach) [reply tests/js/common/shell/shell-pregel-status-writer.js:240]: https://arangodb.atlassian.net/browse/GORDO-1595

File: tests/js/common/shell/shell-pregel-status-writer.js [hunk 14]
```diff
@@ -118,6 +216,124 @@ function pregelStatusWriterSuite() {
...
+        const deleted = pregel.removeHistory();
+        assertTrue(deleted === true);
+        const result = pregel.history();
+        assertTrue(Array.isArray(result));
+        assertEqual(result.length, 0);
+      }
+    },
+
+    testRemoveHistoricPregelEntryHTTP: function () {
+      if (isServer) {
```
[the remark this answers is not in the document]
Comment (hkernbach) [reply tests/js/common/shell/shell-pregel-status-writer.js:293]: https://arangodb.atlassian.net/browse/GORDO-1595

File: tests/js/common/shell/shell-pregel-status-writer.js [hunk 15]
```diff
@@ -55,6 +63,98 @@ function pregelStatusWriterSuite() {
...
+          check = true;
+        }
+      } catch (ignore) {
+      }
+    }
+
+    return check;
+  };
+
+  const verifyPersistedStatePIDRead = (pid, expectedState) => {
```
[the remark this answers is not in the document]
Comment (hkernbach) [reply tests/js/common/shell/shell-pregel-status-writer.js:91]: https://arangodb.atlassian.net/browse/GORDO-1595

File: tests/js/common/shell/shell-pregel-status-writer.js [hunk 16]
```diff
@@ -55,6 +63,98 @@ function pregelStatusWriterSuite() {
...
+      assertEqual(apiPersistedState.data.state, expectedState);
+    }
+
+    // 3.) verify using Pregel JavaScript module
+    const modulePersistedState = pregel.history(pid);
+    assertTrue(modulePersistedState);
+    assertEqual(modulePersistedState.data.state, expectedState);
+  };
+
+  const verifyPersistedStatePIDNotAvailableRead = (pid, testDocumentAPI) => {
```
[the remark this answers is not in the document]
Comment (hkernbach) [reply tests/js/common/shell/shell-pregel-status-writer.js:113]: https://arangodb.atlassian.net/browse/GORDO-1595

File: arangod/Pregel/StatusWriter/CollectionStatusWriter.h [hunk 17]
```diff
@@ -95,7 +139,8 @@ struct CollectionStatusWriter : StatusWriterInterface {
 
  private:
   [[nodiscard]] auto createOperation(OperationType operation,
```
[the remark this answers is not in the document]
Comment (hkernbach) [reply arangod/Pregel/StatusWriter/CollectionStatusWriter.h:141]: https://arangodb.atlassian.net/browse/GORDO-1593

**Pull request merged**